src/run.rs

```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
             output
                 .lines()
                 .map(|line| line.trim().to_owned())
-                .filter(|line| !line.is_empty())
+                .filter_map(|line| {
+                    if line.is_empty() {
+                        let tab_index = line.find('\t').unwrap();
+                        let (image_id, date_str) = line.split_at(tab_index);
+                        Some((image_id.to_owned(), parse_docker_date(date_str.trim())))
+                    } else {
+                        None
+                    }
+                })
                 .collect()
         })
         .map_err(|error| io::Error::new(io::ErrorKind::Other, error))
 }
 
+// Extract a Duration from Docker's non-standard date format used in `docker image ls`
+// Example input: "2017-12-20 16:30:49 -0500 EST"
+fn parse_docker_date(date_str: &str) -> Duration {
+    let date_fmt = "%Y-%m-%d %H:%M:%S %z";
+    // Chrono can't read the "EST", so remove it before parsing
+    let timezone_index = 25;
+
+    let timestamp = if let Ok(dt) = DateTime::parse_from_str(&date_str[..timezone_index], date_fmt)
+    {
+        dt.timestamp()
+    } else {
+        warn!(
+            "Unable to parse CreatedAt: {}. Defaulting to current time",
+            date_str
+        );
+        Local::now().timestamp()
+    };
+    Duration::from_secs(timestamp.try_into().unwrap())
```

Comment:
I have a slight preference for not converting the timestamp into an integer and then creating a duration out of that. The reason is that we lose some type safety: we have to rely on the fact that the integer represents a number of seconds since the UNIX epoch, which is in some sense a lower-level implementation detail of how timestamps are represented.

I think it could be better to be explicit about this calculation like we do here:

https://github.com/stepchowfun/docuum/blob/b246d638c0d99d0f40b06822685cc44dd307e055/src/run.rs#L233

(But instead of `SystemTime::now()`, maybe `Utc::now()` since we are in `chrono`-land here?)

Unfortunately it would be more verbose, though.

More importantly though: I prefer to avoid the `unwrap` if possible (and if not possible, justify it with a comment).

Reply:
Is it possible to convert from a `chrono::DateTime` to a `std::time::Duration` without an intermediate step as `i64`? Looks like the direct conversion from SystemTime -> Duration is unique to SystemTime.

What's the risk if we do that conversion to i64? Wouldn't the risk just be deleting images slightly less optimally if the system time doesn't increase monotonically? It wouldn't crash, would it?

Reply:
For context, this is what I have so far to remove that `unwrap()`, but it still has the type conversion:

```
    let dt = match DateTime::parse_from_str(&clean_date_str, "%Y-%m-%d %H:%M:%S %z") {
        Ok(dt) => dt,
        Err(e) => return Err(io::Error::new(io::ErrorKind::Other, e)),
    };
    let num_seconds = match u64::try_from(dt.timestamp()) {
        Ok(num_seconds) => num_seconds,
        Err(e) => return Err(io::Error::new(io::ErrorKind::Other, e)),
    };
    Duration::from_secs(num_seconds))
```

Reply:
> Is it possible to convert from a `chrono::DateTime` to a `std::time::Duration` without an intermediate step as `i64`? Looks like the direct conversion from SystemTime -> Duration is unique to SystemTime.

I think `chrono::DateTime` has [`signed_duration_since`](https://docs.rs/chrono/0.4.13/chrono/struct.DateTime.html#method.signed_duration_since). Maybe we can use that? I'm a little bit confused though by the fact that it returns an `OldDuration`, and I'm not sure what that is. If my suggestion turns out to be too awkward to implement, we can go with your implementation (from your comment in this thread).

> What's the risk if we do that conversion to i64? Wouldn't the risk just be deleting images slightly less optimally if the system time doesn't increase monotonically? It wouldn't crash, would it?

It's not a correctness issue (I think the non-monotonicity concern applies equally to both cases?), it's just a (very) minor style issue. Going from a `DateTime` directly to a `Duration` seems (to me, at least) slightly better than doing an intermediate conversion to a plain integer, because with an integer we're no longer encoding our intent in the type system. It's a lower-level representation that offers no guarantees about how it's used.

In my mind, it's kind of like creating [newtypes](https://doc.rust-lang.org/stable/rust-by-example/generics/new_types.html) to represent different units like `Meters`, `Miles`, etc. They're all integers under the hood, but using a more specific type helps the type system catch mistake.

Let me know if that makes sense.

Reply:
Just to clarify: I think for this particular code it doesn't matter that much, because the logic is simple enough that it's unlikely there will be a hidden mistake in it. But for more complicated logic it really does help to leverage the type system like this, so I prefer to just apply this kind of rigor to everything, no matter how trivial. But I understand this makes me an annoying code reviewer sometimes!

Reply:
`OldDuration` is an alias for `chrono::Duration`. See https://docs.rs/chrono/0.4.13/chrono/index.html#duration

Reply:
Ahh yeah that works. I remember seeing that `signed_duration_since` function, but skipped it since I couldn't think of a nice way to pass `UNIX_EPOCH` into that function (until I discovered `DateTime::from(UNIX_EPOCH)` today)